src/ModelContextProtocol.Analyzers/XmlToDescriptionGenerator.cs

```diff
@@ -0,0 +1,386 @@
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Text;
+using System.Collections.Immutable;
+using System.Text;
+using System.Xml.Linq;
+
+namespace ModelContextProtocol.Analyzers;
+
+/// <summary>
+/// Source generator that creates [Description] attributes from XML comments
+/// for partial methods tagged with MCP attributes.
+/// </summary>
+[Generator]
+public sealed class XmlToDescriptionGenerator : IIncrementalGenerator
+{
+    private const string GeneratedFileName = "ModelContextProtocol.Descriptions.g.cs";
+
+    public void Initialize(IncrementalGeneratorInitializationContext context)
+    {
+        // Filter method declarations with attributes. We're looking for attributed partial methods.
+        var methodModels = context.SyntaxProvider
+            .CreateSyntaxProvider(
+                static (s, _) => s is MethodDeclarationSyntax { AttributeLists.Count: > 0 } method && method.Modifiers.Any(SyntaxKind.PartialKeyword),
+                static (ctx, ct) =>
+                {
+                    var methodDeclaration = (MethodDeclarationSyntax)ctx.Node;
+                    return ctx.SemanticModel.GetDeclaredSymbol(methodDeclaration, ct) is { } methodSymbol ?
+                        new MethodToGenerate(methodDeclaration, methodSymbol) :
+                        null;
+                })
+            .Where(static m => m is not null);
+
+        // Combine with compilation to get well-known type symbols.
+        var compilationAndMethods = context.CompilationProvider.Combine(methodModels.Collect());
```

Comment:
By using `Combine` on the compilation on the final incremental value, we're effectively telling the source generator to trigger an E2E source generation every time the compilation gets updated. We should try to update the incremental graph such that compilation information is mapped out to a final, structurally equatable incremental value.